### PR TITLE
statistics: correctly check if a column histogram is loaded

### LINKED ISSUE #31925: statistics: column stats are not loaded on-demand if its histogram is empty
## Enhancement
In the following example, we analyze table `t` and query `select * from a = 1`. The stats of column `a` are loaded as expected. Then we insert more values into `t` and analyze it again. Then we find that the stats of column `a` are not loaded. Notice that in the case the histogram of column `a` is empty.
飞书20220107-132615

For the case the histogram of column `a` is not empty, after we analyzing table `t` again, we can see the new stats of column `a` are loaded in memory:
飞书20220107-132630

We hope for the case the histogram is empty, needed column stats can also be loaded into memory after analyzing.

### LINKED ISSUE #32833: avoid using pseudo estimation strategy for fully loaded column stats
## Enhancement
For the column stats, the pseudo estimation strategy is expected to be used when the histograms are not loaded.
But sometimes, the check for whether the histogram is loaded or not is not accurate. 
Currently, it's decided by checking `notNullCount == 0 && NDV > 0`. But it's possible that stats of this column don't have histogram and topn, which would make the checking result wrong. In the following two cases, this might happen:
1. There are many NULLs in this column.
2. The row count of this table is very small.

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #32833  close #31925 ref #32758

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -1108,20 +1112,23 @@ func (c *Column) IsInvalid(sctx sessionctx.Context, collPseudo bool) bool {
 		if stmtctx != nil && stmtctx.StatsLoad.Fallback {
 			return true
 		}
-		if c.Histogram.NDV > 0 && c.notNullCount() == 0 && stmtctx != nil {
+		if !c.Loaded && stmtctx != nil {
```
Comment (qw4990) [inline statistics/histogram.go:1115]: How about wrapping this logic in a dedicated function like `c.IsLoaded()`?
Then we can implement it as `c.Loaded || (c.Histogram.NDV > 0 && c.notNullCount() == 0)` which is safer and easier to maintain(for example, if we want to update this logic in the future, we just need to modify one place).

Comment (time-and-fate) [reply statistics/histogram.go]: Updated.
But I'd like to think of it as insurance as we are switching from the old method to using `Loaded`. I expect to remove the method someday and use the single `Loaded` variable to check.

**Review by xuyifangreeneyes — COMMENTED** [review]
We identify whether column stats are loaded by checking `statistics.Column.Len()`(which is the number of buckets) in some places such as:
https://github.com/pingcap/tidb/blob/284b90d2d50fa7b4991db6e85a8d46087f8cb13a/statistics/handle/handle.go#L787-L796
and
https://github.com/pingcap/tidb/blob/284b90d2d50fa7b4991db6e85a8d46087f8cb13a/statistics/handle/handle.go#L616-L619
Maybe we can also change the check to `IsLoaded()`.
Besides, I think #31925 is also related to wrong check of whether column stats are loaded. Maybe the pr can also fix that.

**Review by qw4990 — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]

**Pull request merged**